### ISSUE TITLE
Reproduction of results in the original paper 

### ISSUE THREAD

**Issue by Harry-Up** [ada-shen/ICNN#12]
Hi, can anyone reproduce the results in the raw paper based on this pytorch project, especially the average part interpretability of filters?

On the one hand, neither the matlab nor the pytorch project provides the relevant metric code. On the other hand, when I want to conduct the multi-classification experiment on the VOC dataset based on this pytorch project, the accuracy is stuck at a low level. Fortunately, the original matlab code works fine in the multi-classification experiments.

Additionally, I am confused when I want to migrate the backbone in the original matlab project from vgg16 to resnet18. Because ICNN turns the last conv  layer and the first fc layer into masked layers. Then, the interpretability metrics are based on the output of the first masked layer. However, it will miss information about skip connections in resnet, so that not all output information is under constraints. Does anyone have an idea to modify the original matlab code?

**Comment by Lixiaomemg** [comment]
Hi, how are 'train.mat', 'val.mat', 'mean.mat', '_train.mat', '_val.mat', '_mean.mat', generated in the get_imdb_multi function in the load_data_multi.py file? What files are they?

**Comment by colorfuldarkgray** [comment]
If I remember correctly those are files containing or stacking all the
samples (intensity normalized) for different partitions.

**Comment by Lixiaomemg** [comment]
How should I get these files?

**Comment by Lixiaomemg** [comment]
Is there a python file for processing the dataset in question?

**Comment by Harry-Up** [comment]
Maybe, you could refer to the official Matlab code for more help, i.e., https://github.com/zqs1022/interpretableCNN. However, last year when I try to run this PyTorch project, I cannot get a well-trained model. I also tried to directly write the loss of the original paper in the form of native Pytorch and automatic derivation, but the training also collapsed.

Unfortunately, the visualization results of the model run from the official Matlab code show that ICNN may have some problems. At the global level, we averaged the activation maps of filters. It is almost a circle for all data, which lacks a diversity of interpretability. At the local level, we exploit the feature map before the EXTRA Gaussian mask layer. It is quite scattered and may not be interpretable.

What's worse, we further explored the quantitative results and they may be some problems. For the average part interpretability of filters, I saved the feature map before the EXTRA mask layer and took advantage of the metric code from https://github.com/hyliang96/CSGCNN/blob/master/VOCPart/activmap/activmap.py. We could not reproduce the result in the original paper, and this is why I raise this issue. For the location instability of filters, we simply changed the calculation of the focus point from the position of peak value to the mean position (\mu of a 2D Gaussian distribution) by L1 or softmax. It seems not better than a normal model or even a model without training.

![1677228918718](https://user-images.githubusercontent.com/26539945/221135412-bf6a2908-ae46-4727-a1b7-8ec105c407c4.jpg)

Eventually, we really appreciate the idea and theoretical derivation in the original paper and believe that they have indeed achieved a good result. However, the official code written by Matlab stopped us from verifying. Unfortunately, this project directly transcribes the back-propagation derivation calculation in the official Matlab code into Pytorch, which may have some bugs. Inspired by ICNN and icCNN, we proposed a novel active interpretability loss and a series of interpretability metrics for hidden-semantic-based interpretability assessment. The results show that the proposed method seems to solve the above problems. It is still under review.

**Comment by colorfuldarkgray** [comment]
It's created from dataset. Sorry, don't remember details.

**Comment by Lixiaomemg** [comment]
That's okay. Thank you.
